cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskManager.java

```diff
@@ -360,169 +213,81 @@ public synchronized void markTaskDone(ExecutionTask task) {
-                          Set<ExecutionTask> inProgressTasks,
-                          Set<ExecutionTask> abortingTasks,
-                          Set<ExecutionTask> abortedTasks,
-                          Set<ExecutionTask> deadTasks,
-                          long remainingDataToMoveInMB) {
-      _remainingPartitionMovements = remainingPartitionMovements;
-      _remainingLeadershipMovements = remainingLeadershipMovements;
-      _inProgressTasks = inProgressTasks;
-      _abortingTasks = abortingTasks;
-      _abortedTasks = abortedTasks;
-      _deadTasks = deadTasks;
-      _remainingDataToMoveInMB = remainingDataToMoveInMB;
-    }
+  public synchronized long finishedDataMovementInMB() {
+    return _executionTaskTracker.finishedDataMovementInMB();
+  }
 
-    /**
-     * Returns the remaining partition movement tasks.
-     */
-    public Set<ExecutionTask> remainingPartitionMovements() {
-      return _remainingPartitionMovements;
-    }
+  public synchronized Set<ExecutionTask> inExecutionTasks() {
+    return inExecutionTasks(TaskType.cachedValues());
+  }
 
-    /**
-     * Returns the remaining leadership movement tasks.
-     */
-    public Collection<ExecutionTask> remainingLeadershipMovements() {
-      return _remainingLeadershipMovements;
-    }
+  public synchronized Set<ExecutionTask> inExecutionTasks(Collection<TaskType> types) {
+    return _executionTaskTracker.inExecutionTasks(types);
+  }
 
-    /**
-     * Get all the in-progress execution tasks.
-     */
-    public Set<ExecutionTask> inProgressTasks() {
-      return _inProgressTasks;
-    }
+  public synchronized long inExecutionDataToMoveInMB() {
+    return _executionTaskTracker.inExecutionDataMovementInMB();
+  }
 
-    /**
-     * @return the aborting tasks.
-     */
-    public Set<ExecutionTask> abortingTasks() {
-      return _abortingTasks;
-    }
+  public synchronized int numRemainingLeadershipMovements() {
+    return _executionTaskTracker.numRemainingLeadershipMovements();
+  }
 
-    /**
-     * @return the aborted tasks.
-     */
-    public Set<ExecutionTask> abortedTasks() {
-      return _abortedTasks;
-    }
+  public synchronized int numFinishedLeadershipMovements() {
+    return _executionTaskTracker.numFinishedLeadershipMovements();
+  }
 
-    /**
-     * @return the dead tasks.
-     */
-    public Set<ExecutionTask> deadTasks() {
-      return _deadTasks;
-    }
+  public synchronized void clear() {
+    _brokersToSkipConcurrencyCheck.clear();
+    _inProgressReplicaMovementsByBrokerId.clear();
+    _inProgressPartitions.clear();
+    _executionTaskPlanner.clear();
+    _executionTaskTracker.clear();
+  }
 
-    /**
-     * Returns the remaining data to move in MB.
-     */
-    public long remainingDataToMoveInMB() {
-      return _remainingDataToMoveInMB;
-    }
+  public synchronized ExecutionTasksSummary getExecutionTasksSummary(List<TaskType> taskTypesToGetSnapshot,
```

Comment:
I am worried that having an _implicit_ filtering based on (1) task types and (2) `inProgressOnly` may cause confusion.

For example, during a replica movement, it will show that there are `N` pending tasks, which will go down to `0` and then will increase again to `M` for leadership movements. This is a silent behavior change that user have no control over.

Rather than a silent change, I feel that it would make it more obvious to the user if we let them control such a filtering on the executor state response using some filtering parameters (if requested so). Hence, we can preserve the existing default response behavior with additional functionality to filter (if requested).

Reply:
The main reason I have a filtering here is `_taskSnapshot` is a large object(if full populated it will contains handle for ExecutionTask objects of all type and state). So I want save some memory here and only store necessary task handles which are need for [response](https://github.com/linkedin/cruise-control/blob/37e77b32bafd288ea81594b9a498b758a71fbd41/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutorState.java#L171). 

And whenever returns to client, the length of task list always match the statistic number( means if non-verbose response says there are N pending replica movement task, then in verbose response there will be N task in pending list) because they are generating while [holding ExecutionTaskManager](https://github.com/linkedin/cruise-control/blob/37e77b32bafd288ea81594b9a498b758a71fbd41/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskManager.java#L289)

Reply:
It would be as before, i.e when executor is executing tasks, the what user get will be a little bit stale, but since both get the value from [_tasksByType](https://github.com/linkedin/cruise-control/blob/90e0e3f4a2e4044718badf9661e90bb18c0ce9a0/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskTracker.java#L26), the value would be consistent(it is like snapshot isolation here...).